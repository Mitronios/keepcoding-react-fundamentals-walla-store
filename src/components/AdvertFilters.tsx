import { useEffect, useState } from "react";
import type { filtersType } from "../interfaces/filtersType";
import { getAvailableTags } from "../services/advertServices";

interface AdvertFiltersProps {
	onChange: (filters: filtersType) => void;
}

const AdvertFilters = ({ onChange }: AdvertFiltersProps) => {
	const [sale, setSale] = useState("all");
	const [tags, setTags] = useState<string[]>([]);
	const [availableTags, setAvailableTags] = useState<string[]>([]);

	useEffect(() => {
		const fetchTags = async () => {
			try {
				const tagsFromServer = await getAvailableTags();
				setAvailableTags(tagsFromServer);
			} catch (error) {
				console.error("Error loading tags", error);
			}
		};
		fetchTags();
	}, []);

	useEffect(() => {
		const filters: filtersType = {};

		if (sale !== "all") {
			filters.sale = sale === "true";
		}

		if (tags.length > 0) {
			filters.tags = tags.join(",");
		}

		onChange(filters);
	}, [sale, tags]);

	const handleOnChange = (
		event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>,
	) => {
		const target = event.target as HTMLInputElement;
		const { name, value, checked } = target;

		switch (name) {
			case "sale":
				setSale(value);
				break;
			case "tags":
				if (checked) {
					setTags((prev) => [...prev, value]);
				} else {
					setTags((prev) => prev.filter((tag) => tag !== value));
				}
				break;
		}
	};

	return (
		<div className="bg-white p-4 rounded-lg shadow-md mb-6 flex flex-col gap-4 md:flex-row md:items-center md justify-center mx-auto max-w-md">
			{/* Filter by tags */}
			<div className="flex flex-col max-w-xs">
				<label className="text-sm font-medium text-gray-700 mb-1">Tags</label>
				<div className="flex flex-wrap gap-2 max-h-40 overflow-y-auto border rounded p-2">
					{availableTags.map((tag) => (
						<label
							key={tag}
							className="inline-flex items-center space-x-2 cursor-pointer"
						>
							<input
								type="checkbox"
								name="tags"
								value={tag}
								checked={tags.includes(tag)}
								onChange={handleOnChange}
								className="form-checkbox"
							/>
							<span className="text-gray-700 text-sm">{tag}</span>
						</label>
					))}
				</div>
			</div>

			{/* Filter by type */}
			<div className="flex flex-col">
				<label
					htmlFor="filter-sale"
					className="text-sm font-medium text-gray-700"
				>
					Type
				</label>
				<select
					id="filter-sale"
					name="sale"
					value={sale}
					onChange={handleOnChange}
					className="border rounded p-2"
					autoCapitalize="off"
				>
					<option value="all">All</option>
					<option value="true">Sale</option>
					<option value="false">Buy</option>
				</select>
			</div>
		</div>
	);
};

export default AdvertFilters;
