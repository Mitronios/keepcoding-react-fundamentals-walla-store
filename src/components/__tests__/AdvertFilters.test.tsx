import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import AdvertFilters from "../AdvertFilters";
import authReducer from "../../store/slices/authSlice";
import advertsReducer from "../../store/slices/advertsSlice";
import tagsReducer from "../../store/slices/tagsSlice";

const createTestStore = () => {
  return configureStore({
    reducer: {
      auth: authReducer,
      adverts: advertsReducer,
      tags: tagsReducer,
    },
    preloadedState: {
      tags: {
        items: ["mobile", "lifestyle", "motor", "work", "laptop"],
        status: "succeeded" as const,
        error: null,
      },
    },
  });
};

describe("AdvertFilters Component", () => {
  it("should match snapshot", () => {
    const store = createTestStore();

    const { container } = render(
      <Provider store={store}>
        <AdvertFilters />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render all filter options", () => {
    const store = createTestStore();

    const { getByText } = render(
      <Provider store={store}>
        <AdvertFilters />
      </Provider>
    );

    expect(getByText("Tags")).toBeInTheDocument();
    expect(getByText("Type")).toBeInTheDocument();
    expect(getByText("Clear")).toBeInTheDocument();

    expect(getByText("mobile")).toBeInTheDocument();
    expect(getByText("lifestyle")).toBeInTheDocument();
    expect(getByText("motor")).toBeInTheDocument();

    expect(getByText("All")).toBeInTheDocument();
    expect(getByText("Sale")).toBeInTheDocument();
    expect(getByText("Buy")).toBeInTheDocument();
  });
});
