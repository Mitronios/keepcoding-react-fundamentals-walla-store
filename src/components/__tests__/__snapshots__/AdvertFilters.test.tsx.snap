// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AdvertFilters Component > should match snapshot 1`] = `
<div>
  <div
    class="bg-white p-4 rounded-lg shadow-md mb-6 flex flex-col gap-4 md:flex-row md:items-center md:justify-center mx-auto max-w-md"
  >
    <div
      class="flex flex-col max-w-xs"
    >
      <label
        class="text-sm font-medium text-gray-700 mb-1"
      >
        Tags
      </label>
      <div
        class="flex flex-wrap gap-2 max-h-40 overflow-y-auto border rounded p-2"
      >
        <label
          class="inline-flex items-center space-x-2 cursor-pointer"
        >
          <input
            class="form-checkbox"
            name="tags"
            type="checkbox"
            value="mobile"
          />
          <span
            class="text-gray-700 text-sm"
          >
            mobile
          </span>
        </label>
        <label
          class="inline-flex items-center space-x-2 cursor-pointer"
        >
          <input
            class="form-checkbox"
            name="tags"
            type="checkbox"
            value="lifestyle"
          />
          <span
            class="text-gray-700 text-sm"
          >
            lifestyle
          </span>
        </label>
        <label
          class="inline-flex items-center space-x-2 cursor-pointer"
        >
          <input
            class="form-checkbox"
            name="tags"
            type="checkbox"
            value="motor"
          />
          <span
            class="text-gray-700 text-sm"
          >
            motor
          </span>
        </label>
        <label
          class="inline-flex items-center space-x-2 cursor-pointer"
        >
          <input
            class="form-checkbox"
            name="tags"
            type="checkbox"
            value="work"
          />
          <span
            class="text-gray-700 text-sm"
          >
            work
          </span>
        </label>
        <label
          class="inline-flex items-center space-x-2 cursor-pointer"
        >
          <input
            class="form-checkbox"
            name="tags"
            type="checkbox"
            value="laptop"
          />
          <span
            class="text-gray-700 text-sm"
          >
            laptop
          </span>
        </label>
      </div>
    </div>
    <div
      class="flex flex-col"
    >
      <label
        class="text-sm font-medium text-gray-700"
        for="filter-sale"
      >
        Type
      </label>
      <select
        autocapitalize="off"
        class="border rounded p-2"
        id="filter-sale"
        name="sale"
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="true"
        >
          Sale
        </option>
        <option
          value="false"
        >
          Buy
        </option>
      </select>
    </div>
    <button
      class="px-4 py-2 bg-gray-500 text-white rounded hover:bg-gray-600 transition"
    >
      Clear
    </button>
  </div>
</div>
`;
